syntax = "proto3";

package glinthawk.protobuf;

message Hey
{
  enum Platform
  {
    AMD64 = 0;
    CUDA = 1;
  }

  string ip = 1;
  uint32 port = 2;
  Platform platform = 3;
}

message InitializeWorker
{
  string model_name = 1;
  uint32 start_layer = 2;
  uint32 end_layer = 3;
  uint32 n_tier_1 = 4;
  uint32 concurrency_pre_att_size_tier_1 = 5;
  uint32 concurrency_att_size_tier_1 = 6;
  uint32 concurrency_post_att_size_tier_1 = 7;
  uint32 concurrency_cls_size_tier_1 = 8;
  uint32 max_context_count_tier_1 = 9;
  uint32 n_tier_2 = 10;
  uint32 concurrency_pre_att_size_tier_2 = 11;
  uint32 concurrency_att_size_tier_2 = 12;
  uint32 concurrency_post_att_size_tier_2 = 13;
  uint32 concurrency_cls_size_tier_2 = 14;
  uint32 max_context_count_tier_2 = 15;
  int8 tier = 16;
  uint8 rank = 17;
  bool randomize = 18;
  bool static_parent = 19;
  string blobstore_uri = 20;
}

message ProcessPrompts { repeated string prompt_ids = 1; }

message PromptCompleted { repeated string prompt_ids = 1; }

message SetRoute
{
  message LayerToAddress
  {
    enum Stage
    {
      PreAttention = 0;
      Attention = 1;
      PostAttention = 2;
      Classification = 3;
    }

    uint32 layer_num = 1;
    Stage stage = 2;
    int8 tier = 3;
    uint8 rank = 4;
    string ip = 5;
    uint32 port = 6;
  }

  uint32 route_id = 1;
  repeated LayerToAddress layer_to_address = 2;
}

message WorkerStats
{
  uint64 states_received = 1;
  uint64 states_sent = 2;
  uint64 states_processed = 3;

  uint64 tokens_processed = 4;
  uint64 tokens_generated = 5;
  uint64 prompts_completed = 6;

  uint64 bytes_sent = 7;
  uint64 bytes_received = 8;
}

message PushDummyPrompts
{
  uint32 count = 1;
  uint32 batch_size = 2;
}

message Prompt
{
  string id = 1;
  uint32 temperature = 2;
  uint32 max_tokens = 3 [ json_name = "max_tokens" ];
  repeated uint32 prompt = 4 [ json_name = "prompt" ];
  repeated uint32 completion = 5 [ json_name = "completion" ];
  string prompt_text = 6 [ json_name = "prompt_text" ];
  string completion_text = 7 [ json_name = "completion_text" ];
  string user_data = 8 [ json_name = "user_data" ];
}

message PushPrompts { repeated Prompt prompts = 1; }

message PushCompletions { repeated Prompt completions = 1; }
